core/remotes/docker/fetcher.go

```diff
@@ -35,10 +37,168 @@ import (
 	"github.com/klauspost/compress/zstd"
 	digest "github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/semaphore"
 )
 
+type BufferPool struct {
+	sync.Mutex
+	sem    *semaphore.Weighted
+	pool   []bytes.Buffer
+	bufCap int64
+}
+
+func NewBufferPool(size int64, bufCap int64) *BufferPool {
+	pool := make([]bytes.Buffer, 0, size)
+	return &BufferPool{
+		sem:    semaphore.NewWeighted(size),
+		pool:   pool,
+		bufCap: bufCap,
+	}
+}
+
+func (p *BufferPool) Get() *bytes.Buffer {
+	p.sem.Acquire(context.Background(), 1)
+	p.Lock()
+	defer p.Unlock()
+
+	if len(p.pool) == 0 {
+		return bytes.NewBuffer(make([]byte, 0, p.bufCap))
+	}
+	buf := p.pool[len(p.pool)-1]
+	p.pool = p.pool[:len(p.pool)-1]
+	return &buf
+}
+
+func (p *BufferPool) Put(buffer *bytes.Buffer) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.pool = append(p.pool, *buffer)
+	p.sem.Release(1)
+}
+
+var ErrClosedPipe = errors.New("bufpipe: read/write on closed pipe")
+
+type pipe struct {
+	cond       *sync.Cond
+	bufPool    *BufferPool
+	buf        *bytes.Buffer
+	rerr, werr error
+}
```

Comment:
qq: can we add some brief comments describing what `pipe` is used for? Will be helpful for others to understand below code :) thanks

Reply:
added a bunch, LMK !